### PR TITLE
fix(DATA-2414_missingColumns_from_Schema)

### PR DESCRIPTION
**Pull request by Irenez753**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -30,6 +30,8 @@ val jacksonVersion = "2.9.9"
 lazy val excludeJpountz = ExclusionRule(organization = "net.jpountz.lz4", name = "lz4")
 lazy val excludeNetty = ExclusionRule(organization = "io.netty", name = "netty")
 lazy val excludeNettyAll = ExclusionRule(organization = "io.netty", name = "netty-all")
+lazy val excludeJbossNetty = ExclusionRule(organization = "org.jboss.netty", name = "netty")
```
Comment (lyogev) [inline build.sbt]: remove

File: build.sbt [hunk 2]
```diff
@@ -76,6 +78,7 @@ dependencyOverrides += "com.fasterxml.jackson.core" % "jackson-databind" % jacks
 dependencyOverrides += "com.fasterxml.jackson.module" % "jackson-module-scala" % jacksonVersion
 dependencyOverrides += "org.apache.avro" %% "avro" % "1.8.2"
 
+//
```
Comment (lyogev) [inline build.sbt]: remove

File: src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala [hunk 3]
```diff
@@ -233,6 +236,23 @@ class HudiOutputWriter(props: Map[String, Object], hudiOutput: Option[Hudi]) ext
   def isOnlyNullColumn(df: DataFrame, name: String): Boolean = {
     df.select(name).filter(col(name).isNotNull).limit(1).count() == 0
   }
+
+  def alignToSchemaColumns(dataframe: DataFrame, previousSchema: Option[StructType]) : DataFrame = {
+    var df = dataframe
+    var fieldsList = ListBuffer[String]()
+    previousSchema match {
+      case Some(sch) => {
+        sch.fields.filter(f => !df.columns.contains(f.name)).foreach(f => fieldsList += f.name)
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala]: df.select(col("*") +: (sch.fields.filter(f => !df.columns.contains(f.name)).foreach(f => lit(null).as(f.name))): _*)

File: src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala [hunk 4]
```diff
@@ -233,6 +236,23 @@ class HudiOutputWriter(props: Map[String, Object], hudiOutput: Option[Hudi]) ext
   def isOnlyNullColumn(df: DataFrame, name: String): Boolean = {
     df.select(name).filter(col(name).isNotNull).limit(1).count() == 0
   }
+
+  def alignToSchemaColumns(dataframe: DataFrame, previousSchema: Option[StructType]) : DataFrame = {
+    var df = dataframe
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala]: no need at all...

File: src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala [hunk 5]
```diff
@@ -233,6 +236,23 @@ class HudiOutputWriter(props: Map[String, Object], hudiOutput: Option[Hudi]) ext
...
+        sch.fields.filter(f => !df.columns.contains(f.name)).foreach(f => fieldsList += f.name)
+        for (name <- fieldsList) {
+          // scalastyle:off null
+          df = df.withColumn(name, lit(null))
+          // scalastyle:on null
+        }
+      }
+      case None =>
+    }
+    df
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala]: move to case none

File: src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala [hunk 6]
```diff
@@ -233,6 +236,25 @@ class HudiOutputWriter(props: Map[String, Object], hudiOutput: Option[Hudi]) ext
   def isOnlyNullColumn(df: DataFrame, name: String): Boolean = {
     df.select(name).filter(col(name).isNotNull).limit(1).count() == 0
   }
-}
+
+  def alignToSchemaColumns(df: DataFrame, previousSchema: Option[StructType], dropNullColumns: Option[Boolean]) : DataFrame = {
+    this.hudiOutputProperties.dropNullColumns match {
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala]: use dropNullColumns

Comment (lyogev) [reply src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala]: alignToPreviousSchema

File: src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala [hunk 7]
```diff
@@ -185,6 +185,9 @@ class HudiOutputWriter(props: Map[String, Object], hudiOutput: Option[Hudi]) ext
       case None => None
     }
 
+    // By default we will add missing columns according to previous schema, if exists
+    df = alignToSchemaColumns(df, previousSchema, this.hudiOutputProperties.dropNullColumns)
+
     // Update the schema
     val schema = StructType(df.schema.fields.flatMap(
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala]: move entire section below to function

Comment (lyogev) [reply src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala]: And split nullable part to a different part of the code outside function

File: src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala [hunk 8]
```diff
@@ -47,7 +47,7 @@ class HudiOutputWriter(props: Map[String, Object], hudiOutput: Option[Hudi]) ext
     log.info(s"Starting to write dataframe to hudi")
 
     // To support schema evolution all fields should be nullable
-    val df = updateSchema(dataFrame)
+    val df = alignToPreviousSchema(dataFrame)
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala]: nullable is before/after this and this only happens if the flag is enabled

File: src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala [hunk 9]
```diff
@@ -184,13 +183,53 @@ class HudiOutputWriter(props: Map[String, Object], hudiOutput: Option[Hudi]) ext
...
+    )
+    dataFrame.sparkSession.createDataFrame(dataFrame.rdd, schema)
+  }
+
+  def isOnlyNullColumn(df: DataFrame, name: String): Boolean = {
+    df.select(name).filter(col(name).isNotNull).limit(1).count() == 0
+  }
+
+  def alignToSchemaColumns(df: DataFrame, previousSchema: Option[StructType]) : DataFrame = {
+    this.hudiOutputProperties.dropNullColumns match {
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala]: change parameter to alignToPreviousSchema

Comment (lyogev) [reply src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala]: remove match from here

File: src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala [hunk 10]
```diff
@@ -184,13 +183,53 @@ class HudiOutputWriter(props: Map[String, Object], hudiOutput: Option[Hudi]) ext
...
+            df.select(col("*") +: missingColumns :_*)
+          }
+          case None => df
+        }
+      }
+      case _ => df
+    }
+  }
+
+  def updateSchema(dataFrame: DataFrame, previousSchema: Option[StructType]): DataFrame = {
```
Comment (lyogev) [inline src/main/scala/com/yotpo/metorikku/output/writers/file/HudiOutputWriter.scala]: removeNullColumns

**Comment by tooptoop4** [comment]
@Irenez753 @lyogev is this a workaround for https://github.com/apache/incubator-hudi/issues/845 ?

**Comment by lyogev** [comment]
@tooptoop4 in a way yes... but when working with hudi and a proper schema registry this doesn't happen.

**Pull request merged**